transaction/core/src/ring_signature/mod.rs

```diff
@@ -22,26 +22,68 @@ mod key_image;
 mod mlsag;
 mod rct_bulletproofs;
 
-lazy_static! {
-    /// Generators (base points) for Pedersen commitments.
-    /// For commitment to amount 'v' with blinding 'b', we want 'C = v*H + b*G' so commitments to zero are signed on G.
-    /// Note: our H is not the same point as the dalek library's default version
-    pub static ref GENERATORS: PedersenGens = PedersenGens {
-            B: hash_to_point(&RistrettoPublic::from(RISTRETTO_BASEPOINT_POINT)),
-            B_blinding: RISTRETTO_BASEPOINT_POINT
-    };
+/// The base point for blinding factors used with all amount commitments
+pub const B_BLINDING: RistrettoPoint = RISTRETTO_BASEPOINT_POINT;
 
+lazy_static! {
     /// Generators (base points) for Bulletproofs.
     /// The `party_capacity` is the maximum number of values in one proof. It should
     /// be at least 2 * MAX_INPUTS + MAX_OUTPUTS, which allows for inputs, pseudo outputs, and outputs.
     pub static ref BP_GENERATORS: BulletproofGens =
         BulletproofGens::new(64, 64);
 }
 
+/// Generators (base points) for Pedersen commitments to amounts.
+///
+/// For commitment to amount 'v' with blinding 'b', we want 'C = v*H + b*G'
+/// so commitments to zero are signed on G, where G is the ristretto basepoint.
+///
+/// Note: our H is not the same point as the dalek library's default version
+///
+/// For amounts, H varies based on the token id.
+pub fn generators(token_id: u32) -> PedersenGens {
+    let mut hasher = Blake2b::new();
+    hasher.update(&HASH_TO_POINT_DOMAIN_TAG);
+
+    // This step xors the token id bytes on top of the "base point" bytes
+    // used prior to the introduction of token ids.
+    //
+    // This ensures:
+    // * The function is constant-time with respect to token id
+    // * The behavior for id 0 is the same as before
+    // * For different id values, the set of B points are orthogonal.
```

Comment:
I haven't been able to find a reference for "orthogonal" base points. I pulled up some definitions of Pedersen commitments and found a couple of good descriptions:

"Given the two generators G and H, no one knows the discrete log for second generator H with respect to first generator G, meaning no one knows a y such that y ∗ G = H"

or 

Computational soundness (binding): if the generators g, h are independently generated, then finding a discrete logarithm relation between them is as hard as breaking the Discrete Log Problem.